columns,+columns options

kfly8 opened the pull request:
add `columns` and `+columns` search options, as same as Teng.

e.g.

``` perl
# columns opts
my $itr = $db->search_joined(user_item => [
    item => {'user_item.item_id' => 'item.id'},
], {
    'item.id' => 1,
}, {
    'columns'  => ['item.id'],
});
$itr->{sql}
# =>
# SELECT `item`.`id`
# FROM `user_item` INNER JOIN `item` ON `user_item`.`item_id` = `item`.`id`
# WHERE (`item`.`id` = ?)
```

``` perl
# +columns opts
my $itr = $db->search_joined(user_item => [
    item => {'user_item.item_id' => 'item.id'},
], {
    'item.id' => 1,
}, {
    '+columns'  => [\'item.id * 10 AS item_num'],
});
$itr->{sql}
# =>
# SELECT `user_item`.`id`, `user_item`.`user_id`, `user_item`.`item_id`, `item`.`id`, `item`.`name_en`, item.id * 10 AS item_num
# FROM `user_item` INNER JOIN `item` ON `user_item`.`item_id` = `item`.`id`
# WHERE (`item`.`id` = ?)

# get  `+columns` by base_table row.
my ($user_item, $item) = $itr->next;
$user_item->item_num # => 10
```

kfly8 committed:
columns,+columns options

Songmu commented on lib/Teng/Plugin/SearchJoined/Iterator.pm:
```diff
@@ -43,9 +43,14 @@ sub _seperate_rows {
-        my $value = $row->[$i++];
+        my $value = $row->[$i];
         my ($table, $column) = split /$name_sep/, $field;
+        if (!$column) {
```
You should branch if the `$field` is string or reference.

kfly8 replied:
updated: 
https://github.com/kfly8/p5-Teng-Plugin-SearchJoined/blob/columns_opts/lib/Teng/Plugin/SearchJoined/Iterator.pm#L48

Songmu commented on lib/Teng/Plugin/SearchJoined/Iterator.pm:
```diff
@@ -43,9 +43,14 @@ sub _seperate_rows {
         my ($table, $column) = split /$name_sep/, $field;
+        if (!$column) {
+            $column = $self->{sth}->{NAME_lc}->[$i];
+            $table  = $self->{table_names}->[0]; # base_table
```
I think it is not good idea using `base_table` as the table name. How about using '_' or so?

kfly8 replied:
Surely it is I think a good idea. .
But, for the following reasons, 
I wanted to remains `base_table`.
- schema definition is needed when passing `new_row_from_hash` 
  - https://github.com/nekokak/p5-Teng/blob/master/lib/Teng.pm#L646-L647
- I want to similar shape and when you use the `+columns` in Teng

e.g.  Teng +columns example

``` perl

use Test::More;
use DBI;
use Teng;
use Teng::Schema::Declare;

my $dbh = DBI->connect('dbi:mysql:test', 'root', '');

my $schema = schema {
    table {
        name "foo";
        columns qw( id );
    };
} "MyDB::Schema";

my $teng = Teng->new(
    dbh    => $dbh,
    schema => $schema,
);

my $row = $teng->single(foo => +{}, { '+columns' => [\'1+1 as hello']});
is $row->hello, 2;

done_testing;
```

Songmu commented on lib/Teng/Plugin/SearchJoined.pm:
```diff
@@ -38,12 +32,34 @@ sub search_joined {
+    }
+
+    if ($opt->{'+columns'}) {
+        push @fields => @{$opt->{'+columns'}}
```
[nit] Could you use comma instead of fat comma?

kfly8 replied:
Sorry.. fixed: https://github.com/kfly8/p5-Teng-Plugin-SearchJoined/blob/columns_opts/lib/Teng/Plugin/SearchJoined.pm#L57

Songmu commented on lib/Teng/Plugin/SearchJoined.pm:
```diff
@@ -22,15 +22,9 @@ sub search_joined {
-        my @columns = map { "$table_name$name_sep$_" } @{ $table->columns };
-        push @fields, @columns;
-    }
+    my $fields = _get_select_joined_columns($self, \@tables, $opt);
```
[nit] How about using `$self->_get_select..`?

kfly8 replied:
updated: https://github.com/kfly8/p5-Teng-Plugin-SearchJoined/blob/columns_opts/lib/Teng/Plugin/SearchJoined.pm#L25

But export `_get_select_joined_columns` : https://github.com/kfly8/p5-Teng-Plugin-SearchJoined/blob/columns_opts/lib/Teng/Plugin/SearchJoined.pm#L12

Is it ok..?

Songmu commented:
Thank you for pull request. Add some comments.

kfly8 committed:
clean

- https://github.com/Songmu/p5-Teng-Plugin-SearchJoined/pull/2#discussion-diff-28204434
- https://github.com/Songmu/p5-Teng-Plugin-SearchJoined/pull/2#discussion-diff-28204460

kfly8 committed:
export _get_select_joined_columns

kfly8 commented:
Thank you for code reviews!
